contracts/access_control/lib.rs

```diff
@@ -25,10 +31,20 @@ mod access_control {
     pub const HAS_ROLE_SELECTOR: [u8; 4] = [0, 0, 0, 3];
     pub const CHECK_ROLE_SELECTOR: [u8; 4] = [0, 0, 0, 5];
 
-    #[ink(storage)]
-    pub struct AccessControl {
+    pub const STORAGE_KEY: u32 = openbrush::storage_unique_key!(Data);
+
+    #[derive(Debug)]
+    #[openbrush::upgradeable_storage(STORAGE_KEY)]
```

Comment:
I'm trying to learn what this macro actually does. So what's the difference between having `#[openbrush::upgradeable_storage(STORAGE_KEY)]` here vs just `#[ink(storage)]` ?

Reply:
In the "original" contract it does not do that much - it places storage struct under specified storage key (it's clear where to migrate from). If it were in the upgraded version, it additionally allows reading/writing to the uninitialized fields (initializes with default on the first use), making migrations straightforward (at least when no Mappings are involved).

Reply:
IMO that's the killer feature - it allows one to implicitly initialize the new upgraded contract

Reply:
Something is wrong here. I have made quite a bit of tests here and even with all these macros, this is still using the standard storage key of `00000000`. My suspicion is that we are using the macro `#[openbrush::upgradeable_storage(STORAGE_KEY)]` incorrectly. (the annoying thing though is that the docs page is down today https://727-ventures.github.io/openbrush-contracts/ so its not possible to consult it). 
Did we ever test what it does, vs what is it supposed to do?

Reply:
The way to confirm this is to inspect the `layout` part of the json metadata file. You can clearly see that `_reserved` uses the key of `00000000` no matter what "hardcoded" value we put as an argument to the macro.

This is also consistent with what lands on chain.